### ISSUE TITLE
Make Authelia act as a standalone authenticating reverse proxy

### ISSUE THREAD

**Issue by clems4ever** [clems4ever/authelia#467]
Up to now Authelia acted as a companion of reverse proxies. Having a reverse proxy brings flexibility as proxies come with many useful features like caching and other things but in many cases the deployment of a reverse proxy AND Authelia to secure a simple web service is too complicated. It brings many question on the configuration of the reverse proxy to make it work. The authentication processes must be simplified to the maximum for such simple use cases.

Ideally, it should be easy to run Authelia as a standalone authenticating reverse proxy to protect the application and provide all the 2FA sweet features provided by Authelia (many 2FA methods, reset password, login portal, etc...).

**Comment by nightah** [comment]
I like the thought process behind this, however, I question it's actual usefulness.

I dare say that most users that host many applications, or even single applications likely will already be utilising a reverse proxy. I feel this may be more effort than it's worth especially from the perspective of support.

Ideally we need to make information and documentation more readily available, perhaps instead of it hiding between many files which are scattered actually utilising the wiki section of GitHub and covering the most common cases for reverse proxies (nginx and traefik) in more detail.

**Comment by clems4ever** [comment]
I agree that we lack documentation and we'll definitely work to make it much better. There is another point that I did not mention in the original description. It is *security*. It brings security to a higher level because by putting Authelia directly in front of your service, you get an extra level of security between the proxy and the service. One example of such proxy if the oauth2 one from bitly: https://github.com/bitly/oauth2_proxy.

In a one node setup (nginx, Authelia and the service being on the same machine) it might not be very useful but in a complex environment it's definitely something to consider. This relates to the topic of tampering information provided by Authelia to the backend like the forwarded user and forwarded groups. An attacker within the network could fake a user or groups by sending crafted requests to the service, Authelia does not (yet) protect against that. One way to fix it would be to sign (and even encrypt to prevent user enumeration by a MITM) the communication between Authelia and the service by either making the app verify a signature coming along with the headers (this requires the app to be modified if it does not support it though...) or add a daemon doing so in front of the service. In the latter case, it's Authelia or a lite version of it acting a signing reverse proxy.

I will instead relabel the issue into: "Secure information exchange between Authelia and applications". What do you think? Is it clear enough?

**Comment by Rihan9** [comment]
Hi,
I'm not very favorable to this, because I think Authalia should be the security layer, and the reverse proxy should be the reverse proxy. The roles shouldn't be mixed. Configure and run a Caddy reverse proxy is pretty easy and is well integrated with authalia, so I don't really see a reason why authalia should be reinventing the wheel and add another feature to maintain
